### ISSUE TITLE
IteratorAggregate

### ISSUE THREAD

**Issue by natebrunette** [morrisonlevi/php-generics-design-library#2]
https://github.com/morrisonlevi/php-generics-design-library/blob/728fde4705f42c7050a8c7da7011cecc83fe8fbf/src/interfaces.php#L5-L6

This would make more sense to me as

```php
interface IteratorAggregate<out K = mixed, out V = mixed, out It: Traversable<K, V> = Traversable> extends Traversable<K, V> 
```

**Comment by morrisonlevi** [comment]
I just had a horrible realization. `IteratorAggregate::getIterator()` currently has no return type; e.g.:

```php
interface IteratorAggregate {
  function getIterator();
}
```
I already knew this, of course. However, I did not grasp all the implications.

It must accept a lack of a return type to be compatible; if we added a return type that is upper bounded by `Traversable` it will break code like this:

```php
class Vector implements IteratorAggregate {
  function getIterator() {
      return new EmptyIterator();
  }
}
```

Also, this works without warnings of any kind on PHP 7:

```php
class Vector implements IteratorAggregate {
  function getIterator(): string {
      return "string";
  }
}
$vec = new Vector();
var_dump($vec->getIterator());
```

I thought the engine checked for Traversable, so this would/could be an error. Turns out that only certain features like `foreach` check it:

```php
// this throws an exception
foreach ($vec as $key => $value) {
    var_dump($key, $value);
}
```

I wish I'd realized this earlier. However, since people will do things like `getIterator(): VectorIterator` we can't require it to be `Traversable`,`Iterator`, or `Generator` because we don't have covariance yet. 

This means getting covariance into a 7.4, and that we have a 7.4 before 8.0, is very important for generics.

PHP :(

**Comment by natebrunette** [comment]
I see your point. We need covariance and a return type on getIterator before we can make this interface generic. Or we have to make it default to mixed with no upper bound.

**Comment by morrisonlevi** [comment]
> Or we have to make it default to mixed with no upper bound.

This does give some value, but it's limited because you can't correlate it to `Traversable` of particular types.

**Comment by morrisonlevi** [comment]
Oops, did not mean to close the issue; I was just switching windows and happened to click on that exact location.

**Comment by natebrunette** [comment]
Agreed, but it might have to be an intermediary step if something gets stalled.
